pkg/controller/networkpolicy/validate.go

```diff
@@ -119,95 +182,72 @@ func (v *NetworkPolicyValidator) Validate(ar *admv1.AdmissionReview) *admv1.Admi
+			if !allowed {
+				return reason, allowed
+			}
 		}
-		if ok := v.tierExists(tier); !ok {
-			allowed = false
-			reason = fmt.Sprintf("tier %s does not exist", tier)
+	case admv1.Update:
+		for _, val := range v.antreaPolicyValidators {
+			reason, allowed = val.updateValidate(curObj, oldObj, userInfo)
+			if !allowed {
+				return reason, allowed
+			}
 		}
 	case admv1.Delete:
-		// Delete of Antrea Policies have no validation
-		allowed = true
-	}
-	return reason, allowed
-}
-
-// validateRuleName validates if the name of each rule is unique within a policy
-func (v *NetworkPolicyValidator) validateRuleName(ingress, egress []secv1alpha1.Rule) bool {
-	uniqueRuleName := sets.NewString()
-	isUnique := func(rules []secv1alpha1.Rule) bool {
-		for _, rule := range rules {
-			if uniqueRuleName.Has(rule.Name) {
-				return false
+		// Delete of Antrea Policies have no validation. This will be an
+		// empty for loop.
+		for _, val := range v.antreaPolicyValidators {
+			reason, allowed = val.deleteValidate(curObj, userInfo)
+			if !allowed {
+				return reason, allowed
 			}
-			uniqueRuleName.Insert(rule.Name)
 		}
-		return true
 	}
-	return isUnique(ingress) && isUnique(egress)
+	return reason, allowed
 }
 
 // validateTier validates the admission of a Tier resource
-func (v *NetworkPolicyValidator) validateTier(curTier, oldTier *secv1alpha1.Tier, op admv1.Operation) (string, bool) {
+func (v *NetworkPolicyValidator) validateTier(curTier, oldTier *secv1alpha1.Tier, op admv1.Operation, userInfo authenticationv1.UserInfo) (string, bool) {
 	allowed := true
 	reason := ""
 	switch op {
 	case admv1.Create:
 		klog.V(2).Info("Validating CREATE request for Tier")
-		if len(v.networkPolicyController.tierInformer.Informer().GetIndexer().ListIndexFuncValues(PriorityIndex)) >= maxSupportedTiers {
-			return fmt.Sprintf("maximum number of Tiers supported: %d", maxSupportedTiers), false
-		}
-		// Tier priority must not overlap reserved tier's priority
-		if reservedTierPriorities.Has(curTier.Spec.Priority) {
-			return fmt.Sprintf("tier %s priority %d is reserved", curTier.Name, curTier.Spec.Priority), false
-		}
-		// Tier priority must not overlap existing tier's priority
-		trs, err := v.networkPolicyController.tierInformer.Informer().GetIndexer().ByIndex(PriorityIndex, strconv.FormatInt(int64(curTier.Spec.Priority), 10))
-		if err != nil || len(trs) > 0 {
-			return fmt.Sprintf("tier %s priority %d overlaps with existing Tier", curTier.Name, curTier.Spec.Priority), false
+		for _, val := range v.tierValidators {
+			reason, allowed = val.createValidate(curTier, userInfo)
+			if !allowed {
+				return reason, allowed
+			}
 		}
 	case admv1.Update:
 		// Tier priority updates are not allowed
 		klog.V(2).Info("Validating UPDATE request for Tier")
-		if curTier.Spec.Priority != oldTier.Spec.Priority {
-			allowed = false
-			reason = "update to Tier priority is not allowed"
+		for _, val := range v.antreaPolicyValidators {
```

Comment:
should it be `v.tierValidators`?

Reply:
this was to test whether the e2e tests really work. glad they failed. ok I'm lying it was copy paste error. fixed